pkg/util/cloudfoundry/types.go

```diff
@@ -0,0 +1,204 @@
+	ProcessGUID     string
+}
+
+// ActualLRPFromBBSModel creates a new ActualLRP from BBS's ActualLRP model
+func ActualLRPFromBBSModel(bbsLRP *models.ActualLRP) ActualLRP {
+	ports := []uint32{}
+	for _, pm := range bbsLRP.Ports {
+		ports = append(ports, pm.ContainerPort)
+	}
+	a := ActualLRP{
+		AppGUID:     appGUIDFromProcessGUID(bbsLRP.ProcessGuid),
+		CellID:      bbsLRP.CellId,
+		ContainerIP: bbsLRP.InstanceAddress,
+		Index:       bbsLRP.Index,
+		Ports:       ports,
+		ProcessGUID: bbsLRP.ProcessGuid,
+		State:       bbsLRP.State,
+	}
+	return a
+}
+
+// DesiredLRPFromBBSModel creates a new DesiredLRP from BBS's DesiredLRP model
+func DesiredLRPFromBBSModel(bbsLRP *models.DesiredLRP) DesiredLRP {
+	envAD := ADConfig{}
+	envVS := map[string][]byte{}
+	actionEnvs := [][]*models.EnvironmentVariable{}
+	appGUID := appGUIDFromProcessGUID(bbsLRP.ProcessGuid)
+	// Actions are a nested structure, e.g parallel action might contain two serial actions etc
+	// We go through all actions breadth-first and record environment from all run actions,
+	// since these are the ones we need to find
+	actionQueue := []*models.Action{bbsLRP.Action}
+	for len(actionQueue) > 0 {
+		action := actionQueue[0]
+		actionQueue = actionQueue[1:]
+
+		if a := action.GetRunAction(); a != nil {
+			actionEnvs = append(actionEnvs, a.Env)
+		} else if a := action.GetTimeoutAction(); a != nil {
+			actionQueue = append(actionQueue, a.Action)
+		} else if a := action.GetEmitProgressAction(); a != nil {
+			actionQueue = append(actionQueue, a.Action)
+		} else if a := action.GetTryAction(); a != nil {
+			actionQueue = append(actionQueue, a.Action)
+		} else if a := action.GetParallelAction(); a != nil {
+			actionQueue = append(actionQueue, a.Actions...)
+		} else if a := action.GetSerialAction(); a != nil {
+			actionQueue = append(actionQueue, a.Actions...)
+		} else if a := action.GetCodependentAction(); a != nil {
+			actionQueue = append(actionQueue, a.Actions...)
+		}
+	}
+
+	var err error
+	for _, envVars := range actionEnvs {
+		for _, ev := range envVars {
+			if ev.Name == EnvAdVariableName {
+				err = json.Unmarshal([]byte(ev.Value), &envAD)
+				if err != nil {
+					log.Errorf("Failed unmarshalling %s env variable for app %s: %s",
+						EnvAdVariableName, appGUID, err.Error())
+				}
+			} else if ev.Name == EnvVcapServicesVariableName {
+				envVS, err = getVcapServicesMap(ev.Value, appGUID)
+				if err != nil {
+					log.Errorf("Failed unmarshalling %s env variable for app %s: %s",
+						EnvVcapServicesVariableName, appGUID, err.Error())
+				}
+			}
+		}
+		if len(envAD) > 0 {
+			// TODO: find out if there might be more different AD env variables in all actions (I think not)
+			break
+		}
+	}
+	d := DesiredLRP{
+		AppGUID:         appGUID,
+		EnvAD:           envAD,
+		EnvVcapServices: envVS,
+		ProcessGUID:     bbsLRP.ProcessGuid,
+	}
+	return d
+}
+
+func appGUIDFromProcessGUID(processGUID string) string {
+	return processGUID[0:36]
```

Comment:
why 36?

Reply:
Because the length of App GUID in cloudfoundry is always 36. As noted when we last talked, we'll actually have to work on this further in a followup PR and will be getting the App GUID in a different way, so this function will go away.